core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,201 @@
+    }
+    effects = effects.withMesosCall(Mesos.Call.Accept(offer.offerId, operations = operations))
+
+    effects
+  }
+
+  private[core] def pendingLaunch(goal: Goal, podRecord: Option[PodRecord]): Boolean = {
+    (goal == Goal.Running) && podRecord.isEmpty
+  }
+
+  private[core] def computeNextStateForPods(frame: Frame)(changedPodIds: Set[PodId]): FrameEffects = {
+    changedPodIds.foldLeft(FrameEffects.empty) { (initialEffects, podId) =>
+      var effects = initialEffects
+
+      frame.podSpecs.get(podId) match {
+        case None =>
+          // TODO - this should be spurious if the podStatus is non-terminal
+          if (frame.podStatuses.get(podId).exists { status =>
+              terminalOrUnreachable(status)
+            }) {
+            effects = effects.withPodStatus(podId, None)
+          }
+
+          if (frame.podRecords.contains(podId)) {
+            // delete podRecord
+            effects = effects.withPodRecord(podId, None)
+          }
+
+        case Some(podSpec) =>
+          if (podSpec.goal == Goal.Terminal) {
+            taskIdsFor(podSpec).foreach { taskId =>
+              effects = effects.withMesosCall(Mesos.Call.Kill(taskId))
+            }
+          }
+      }
+      effects
+    }
+  }
+
+  case class HandleMesosEventResult(frameEffects: FrameEffects, dirtyPodIds: Set[PodId])
+
+  object HandleMesosEventResult {
+    val empty = HandleMesosEventResult(FrameEffects.empty, Set.empty)
+  }
+
+  def handleMesosEvent(frame: Frame, pendingLaunch: Set[PodId])(event: Mesos.Event): HandleMesosEventResult = {
+    event match {
+      case offer: Mesos.Event.Offer =>
+        val changedPodIds = pendingLaunch
+        val effects = matchOffer(offer, pendingLaunch.flatMap { podId =>
+          frame.podSpecs.get(podId)
+        }(collection.breakOut))
+        HandleMesosEventResult(effects, changedPodIds)
+
+      case Mesos.Event.StatusUpdate(taskId, taskStatus) =>
+        val podId = podIdFor(taskId)
+
+        if (frame.podSpecs.contains(podId)) {
+          val newState = frame.podStatuses.get(podId) match {
+            case Some(status) =>
+              status.copy(taskStatuses = status.taskStatuses.updated(taskId, taskStatus))
+            case None =>
+              PodStatus(podId, Map(taskId -> taskStatus))
+          }
+
+          val effects = FrameEffects.empty
+            .withPodStatus(podId, Some(newState))
+
+          HandleMesosEventResult(effects, Set(podId))
+        } else {
+          HandleMesosEventResult.empty
+        }
+    }
+  }
+
+  /**
+    * We remove a task if it is not reachable and running, and it has no podSpec defined
+    *
+    * Should be called with the effects already applied for the specified podIds
+    *
+    * @param podIds podIds changed during the last frame
+    * @return
+    */
+  def pruneTaskStatuses(frame: Frame)(podIds: Set[PodId]): FrameEffects = {
```

Comment:
Hm, it really feels we should return a `Frame` instead.

Reply:
Preference for FrameEffects. See FrameWithEffects abstraction.